File: uvaonlinejudge/00514 - Rails.c
#include <stdio.h>
#define MAX_SIZE 1000

int permutation[MAX_SIZE], stack[MAX_SIZE];

int read_permutation(int n, int permutation[]){
    int i;
    for(i = 0; i < n; i++){
        scanf("%d", &permutation[i]);
        if(i == 0 && permutation[0] == 0)
            return 0;
    }
    return 1;
}

int is_possible(int n, int permutation[], int stack[]){
    int i, coach = 1, stack_size = 0;
    for(i = 0; i < n; i++){
        while(1){
            if(stack_size > 0 && stack[stack_size - 1] == permutation[i]){
                stack_size--;
                break;
            }
            if(coach <= n){
                stack[stack_size++] = coach;
                coach++;
            }else{
                return 0;
            }
        }
    }
    return 1;
}

int main(){
    int n;
    while(scanf("%d", &n) && n){
        while(read_permutation(n, permutation)){
            printf("%s\n", is_possible(n, permutation, stack) ? "Yes" : "No");
        }
        printf("\n");
    }
    return 0;
}

File: uvaonlinejudge/00514 - Rails.py
def read_train():
    return int(input())

def read_permutation():
    line = input()
    if line[0] == '0':
        return None
    return [int(i) for i in line.split()]

def is_possible(n, permutation):
    i, stack = 1,  []
    for j in permutation:
        while True:
            if len(stack) > 0 and stack[-1] == j:
                stack.pop()
                break
            if i <= n:
                stack.append(i)
                i += 1
            else:
                return False
    return True

if __name__ == '__main__':
    for n in iter(read_train, 0):
        for permutation in iter(read_permutation, None):
            print('Yes' if is_possible(n, permutation) else 'No')
        print()
